runtime/obj.c

```diff
@@ -80,19 +81,54 @@ CAMLprim value caml_obj_make_forward (value blk, value fwd)
   return Val_unit;
 }
 
+static struct custom_operations default_custom_ops = {
+  "_default_custom_ops",
+  custom_finalize_default,
+  custom_compare_default,
+  custom_hash_default,
+  custom_serialize_default,
+  custom_deserialize_default,
+  custom_compare_ext_default,
+  custom_fixed_length_default
+};
+
 /* [size] is a value encoding a number of blocks */
 CAMLprim value caml_obj_block(value tag, value size)
 {
   value res;
-  mlsize_t sz, i;
+  mlsize_t sz;
   tag_t tg;
 
   sz = Long_val(size);
   tg = Long_val(tag);
-  if (sz == 0) return Atom(tg);
+
+  /* When [tg < No_scan_tag], [caml_alloc] returns an object whose fields are
+   * initialised to [Val_unit]. Otherwise, the fields are uninitialised. We aim
+   * to avoid inconsistent states in other cases.
+   *
+   * For [Abstract_tag], [Double_tag] and [Double_array_tag], the initial
+   * content is irrelevant.
+   *
+   * For [String_tag], the initial contents do no matter. However, the length
+   * of the string is encoded using the last byte of the block. For this
+   * reason, the blocks with [String_tag] cannot be of size [0]. We initialise
+   * the last byte such that the length returned by [String.length] and
+   * [Bytes.length] is non-negative number.
+   *
+   * For [Custom_tag], the first field stores the block containing custom
+   * operations for hashing, marshalling and polymorphic comparison. For this
+   * reason, [Custom_tag] blocks cannot be of size [0]. We initialise the first
+   * field to a block containing default custom operations.
+   */
   res = caml_alloc(sz, tg);
-  for (i = 0; i < sz; i++)
-    Field(res, i) = Val_long(0);
+
+  if (tg == String_tag) {
+    if (sz == 0) caml_invalid_argument ("Obj.new_block");
+    Field (res, sz - 1) = 1;
+  } else if (tg == Custom_tag) {
+    if (sz == 0) caml_invalid_argument ("Obj.new_block");
+    Custom_ops_val (res) = &default_custom_ops;
```

Comment:
Thanks for the proper initializations of the "ops" field.  Currently, the following segfaults
```
ignore (Obj.new_block 255 1000); Gc.full_major()
```
and now it will not.  I'm still wondering whether it makes any sense at all to create a custom block via `Obj.new_block`, since it's almost impossible to set its "ops" field from OCaml.  The alternative here would be to reject (invalid argument) the case `tg == Custom_tag` and document the limitation.

Reply:
I'm not convinced that we should reject custom block allocations. Here is an example of initializing custom blocks from OCaml using a block of custom operations allocated in C.

```c
/* cstub.c */
#include "caml/mlvalues.h"
#include "caml/custom.h"

int compare (value v1, value v2)
{
  return Field(v1,1) - Field(v2,1);
}

static struct custom_operations custom_ops = {
  "_default_ops",
  custom_finalize_default,
  &compare,
  custom_hash_default,
  custom_serialize_default,
  custom_deserialize_default,
  custom_compare_ext_default,
  custom_fixed_length_default
};

value get_custom_ops (value unit)
{
  return (value)&custom_ops;
}
```

```ocaml
/* test.ml */
external get_custom_ops : unit -> Obj.t = "get_custom_ops"

let custom_ops = get_custom_ops ()

let mk_custom_obj v =
  let c = Obj.new_block 255 2 in
  Obj.set_field c 0 custom_ops;
  Obj.set_field c 1 v;
  c

let c1 = mk_custom_obj (Obj.repr 0)
let c2 = mk_custom_obj (Obj.repr 1)

let _ =
  if c1 = c1 then print_endline "c1 = c1";
  if c2 = c2 then print_endline "c2 = c2";
  if c1 <> c2 then print_endline "c1 <> c2"
``` 

```bash
$ ocamlopt.opt test.ml cstub.c
$ ./a.out
c1 = c1
c2 = c2
c1 <> c2
```

There might be better ways to write this. But I can't convince myself to reject this code.

Reply:
This is only tangentially related to this PR, but have been wishing for a while that we had an OCaml-side API to alloc custom block values. Even if custom block operations remain in C-land (because it's the easiest way to guarantee that they don't rely on the runtime system in unreasonable ways?), it would be nice to be able to create new custom values from OCaml. (Preferably with a proper API rather than `set_field`.)

Reply:
I thought about that kind of code and others before writing my message, but they all have problems.  In your example, `get_custom_ops` returns a naked pointer, and the first `Obj.set_field` replaces a naked pointer by another.  This will be outlawed pretty soon.  Also, since you must write a C primitive anyway, it would be safer, and even nicer to read, to do the allocation on the C side.

Reply:
Another bizarre thing that happens when a custom block is created with `Obj.new_block` instead of with `caml_alloc_custom`: if the block is small enough to be allocated in the minor heap, it will never be finalized, even if the ops are changed later to a `struct custom_operation` that has a finalization function.  (Whether to record minor heap blocks for finalization later is decided at allocation time.)

Reply:
Agreed. The latter behavior is quite bad. I can't see a safe use for `Custom_tag` objects allocated through new_block. I'm happy to reject this case and make the change as you had originally proposed. Given that we're changing the semantics of `new_block`, documenting this in `Changes` is useful. I'll make these changes now.

Reply:
Done.